providers/base/bin/zapper_keyboard_test.py

```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env python3
+"""
+This program tests whether the system responds correctly to an external
+USB keyboard.
+
+To run the test you need Zapper board connected and set up.
+"""
+import os
+import struct
+import sys
+import threading
+from enum import Enum
+
+from checkbox_support.scripts.zapper_proxy import zapper_run  # noqa: E402
+
+ZAPPER_KBD = "usb-Canonical_Zapper_main_board_123456-event-kbd"
+EVENT_BIN_FORMAT = "llHHI"
+
+ROBOT_TESTCASE_COMBO = """
+*** Settings ***
+Library    libraries/ZapperHid.py
+
+*** Test Cases ***
+Press and Release a combination of keys.
+    ${keys}=    Create List     A   B
+    Keys Combo    ${keys}
+"""
+
+ROBOT_TESTCASE_TYPE = """
+*** Settings ***
+Library    libraries/ZapperHid.py
+
+*** Test Cases ***
+Type the requested string
+    Type String    hello
+"""
+
+
+class KeyEvent(Enum):
+    UP = 0
+    DOWN = 1
+
+
+def _listen_keyboard_events(event_file_path, callback):
+    """Listen for keyboard events on the given file."""
+    with open(event_file_path, "rb") as event:
+        while True:
+            read_keyboard_events(event, callback)
+
+
+def read_keyboard_events(event, callback):
+    """Read keyboard events from the given file and run the callback."""
+    data = event.read(struct.calcsize(EVENT_BIN_FORMAT))
+    _, _, event_type, code, value = struct.unpack(EVENT_BIN_FORMAT, data)
+    if event_type == 1:  # 0x01 is for _kbd_ events
+        callback((KeyEvent(value), code))
+
+
+def assert_key_combo(host, events):
+    """Request to press a key combination and expect the corresponding events."""
+
+    events.clear()
+    zapper_run(host, "robot_run", ROBOT_TESTCASE_COMBO.encode(), {}, {})
+
+    assert events == [
+        (KeyEvent.DOWN, 30),
+        (KeyEvent.DOWN, 48),
+        (KeyEvent.UP, 48),
+        (KeyEvent.UP, 30),
+    ]
+
+
+def assert_type_string(host, events):
+    """Request to type a string and expect the corresponding events."""
+
+    events.clear()
+    zapper_run(host, "robot_run", ROBOT_TESTCASE_TYPE.encode(), {}, {})
+
+    assert events == [
+        (KeyEvent.DOWN, 35),
+        (KeyEvent.UP, 35),
+        (KeyEvent.DOWN, 18),
+        (KeyEvent.UP, 18),
+        (KeyEvent.DOWN, 38),
+        (KeyEvent.UP, 38),
+        (KeyEvent.DOWN, 38),
+        (KeyEvent.UP, 38),
+        (KeyEvent.DOWN, 24),
+        (KeyEvent.UP, 24),
+    ]
+
+
+def main(argv):
+    if len(argv) != 2:
+        raise SystemExit("Usage: {} <zapper-ip>".format(argv[0]))
+
+    zapper_kbd_path = f"/dev/input/by-id/{ZAPPER_KBD}"
+    if os.path.exists(zapper_kbd_path):
+        event_file = os.path.realpath(zapper_kbd_path)
+    else:
+        raise SystemExit("Cannot find Zapper Keyboard.")
+
+    events = []
+    threading.Thread(
+        target=_listen_keyboard_events, args=(event_file, events.append), daemon=True
+    ).start()
```

Comment:
Since there's no explicit finish to this thing I'm guessing you count on the thread being killed when the process finishes. That's a bit like not calling `free()` after `malloc()` because it'll get freed regardless.

I'm guessing you've probably considered subclassing thread and making the listening thread **object** handle this, right? Why have you decided to do this anonymous thread object?

where I'm getting at is that the ListenerThread could have a property like `keep_running` that could be used by the infinite[-ish] loop above. This way, in `main` you could set this to false, and the thread would finish in a graceful way.

Reply:
The main blocker here was not the named thread as you imagined but the blocking `read()` function. In https://github.com/canonical/checkbox/pull/707/commits/f5f1d4c8f9201dcc3132b470d3f1d30d5e683140 + https://github.com/canonical/checkbox/pull/707/commits/9d239cd978b1e8ff967f23267841e9c11e16661c I created a new listener class as you suggested and took actions to make the _reading_ section non blocking.

Reply:
Blocker? TBH my cache is thoroughly flushed at this point. But my comment was more about encapsulating the condition on which the thread continues to read() (even if it's blocking). 
It looks great now.